### ISSUE TITLE
Add automatic feature embeddings from list of NetworkX attributes

### ISSUE THREAD

**Issue by ntrpnr** [stellargraph/stellargraph#1047]
### Description

Please consider implementing functionality in the method **from_network** which makes it possible to automatically generate feature vector embeddings based on a list of attributes available in my NetworkX graph.

This method should support heterogeneous graphs with different attribute sets.

### User Story

**As** an ML beginner

**I want to** be able to easily specify a set of attributes on my NetworkX graph which will be used by the from_networkx method to automatically generate feature vector embeddings

**So that** I can use StellarGraph without deep knowledge of Python and ML

### Done Checklist

- [ ] Produced code for required functionality
- [ ] Tests written and coverage checked
- [ ] Code review performed
- [ ] Documentation on Google Docs  (if applicable)
- [ ] Documentation in repo
- [ ] Version number reflects new status
- [ ] CHANGELOG.md updated
- [ ] Team demo

**Comment by huonw** [comment]
Hi @ntrpnr, thanks for filing an issue! We're excited to help you.

Could you give a bit more specifics about the graphs you might be working with? For example, what are the values in the attributes? If you have an example graph that you're happy to share, that would be great too, e.g. `nx.write_gpickle(your_graph, "some/path.pickle")`. This will help us understand exactly the problem to solve.

In order to make progress immediately, I'm going to guess at what you might be doing. I could be wrong, in which case we can ignore it :smile: Assuming you've got functions to transform the features of a particular node type, for a few different node types, I think this is possible by iterating over the graph manually. One way to do this would be to add a `features` attribute to each one, but there's others too (e.g. one could build a Pandas DataFrame for each type):

```python
# fill in these with the appropriate code
def features_for_a(x, y): return np.array([...])
def features_for_b(r, s, t): return np.array([...])

for id, data in nx_graph.nodes(data=True):
    node_type = data["label"]

    if node_type == "a":
        data["features"] = features_for_a(data["x"], data["y"])
    elif node_type == "b":
        data["features"] = features_for_b(data["r"], data["s"], data["t"])
    else:
        raise ValueError(f"unknown node type {node_type!r}")

sg_graph = StellarGraph(nx_graph, node_features="features")
```

That code assumes that the node type is saved in the `label` attribute, and that all `a` nodes have `x` and `y` features, and similarly `b` nodes have `r`, `s` and `t`. If that's not quite right, there's various ways to make it more flexible too.

**Comment by ntrpnr** [comment]
Hi @huonw,

Thank you for your quick and comprehensive answer.
I still think my suggested feature would make a nice addition to StellarGraph as it would be much easier to embed features if StellarGraph could just use the attributes available on a NetworkX Graph.

I've managed to generate the feature embeddings.

I can't share my graph but I can share what it looks like as a StellarGraph:
```StellarDiGraph: Directed multigraph
 Nodes: 495, Edges: 6458

 Node types:
  computer: [5]
    Edge types: computer-default->session
  process: [173]
    Edge types: process-default->process
  session: [310]
    Edge types: session-default->process
  tenant: [2]
    Edge types: tenant-default->computer, tenant-default->user
  user: [5]
    Edge types: user-default->session

 Edge types:
    computer-default->session: [310]
    process-default->process: [185]
    session-default->process: [5643]
    tenant-default->computer: [5]
    tenant-default->user: [5]
    user-default->session: [310]
```

This is how I embed the features:

```
process_feature_names = ['name', 'path', 'path_class', 'file_extension']

df_grouped = df.groupby(['process', 'name', 'path', 'path_class', 'file_extension']).size().reset_index(name='count')

feature_encoding = feature_extraction.DictVectorizer(sparse=False, dtype=float, )
process_features_transformed = feature_encoding.fit_transform(df_grouped[process_feature_names].to_dict('records'))
print(process_features_transformed.shape)
> (173, 291)
```
```
process_features = pd.DataFrame(
    process_features_transformed, index=df_grouped.process, dtype="float64"
)
```

```
features = process_features.to_dict('index')
formatted_features = dict()
```

```
import numpy as np

for node, attribute_dict in features.items():

    formatted_features[f'{node}'] = dict()
    values = np.array(list(attribute_dict.values()))
    formatted_features[f'{node}']['feature'] = values

nx.set_node_attributes(G, formatted_features)
```

```
Gs = StellarGraph.from_networkx(G, node_type_attr='nodetype', node_features="feature")
```

I believe I got it right. 
Now I will try to use this for anomaly detection of processes.

**Comment by huonw** [comment]
> I still think my suggested feature would make a nice addition to StellarGraph as it would be much easier to embed features if StellarGraph could just use the attributes available on a NetworkX Graph.

Sorry, I wasn't meaning to suggest that what I discussed above was enough, just providing some tips to get you started right now, before we resolve this issue to everyone's satisfaction. 😄 

> This is how I embed the features:

Thanks for getting back quickly with so much information and code, that's really helpful for understanding what you're up to!

I made an example notebook based on that, so that I can properly experiment with the code: https://gist.github.com/huonw/833a94fdf0a599947b91ffffc6cf1ce5 (rendered more nicely at https://nbviewer.jupyter.org/gist/huonw/833a94fdf0a599947b91ffffc6cf1ce5 too; that's using a in-development version of StellarGraph and so the output is slightly more informative than 0.10.0, but the functionality is the same). I've also made it slightly simpler using [`pandas.get_dummies`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html) instead of having to do the input preparation for `DictVectorizer`, and using [`iterrows`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html) to add the information into the graph.

As part of making that I noticed that the other common method of adding features (passing a dataframe into `node_features`) doesn't work nicely for this sort of graph, where it seems only one node has features. I opened #1059 to cover generalising `from_networkx` to allow an even simpler form, where we don't have to do the `iterrows` loop or `nx.set_node_attributes` (and preparation):

```python
sg.StellarGraph.from_networkx(
    G,
    node_type_attr="nodetype",
    node_features={"process": process_features_transformed},
)
```

Now, some more questions (sorry!): is `df` a dataframe of information you've read from a file like a CSV file? How do you create the NetworkX graph? This will help us understand a bit about where the data is coming from and so what's useful for preparation (both in the form of new features and better documentation).

> Now I will try to use this for anomaly detection of processes.

Nice! If you get something working, we'd love to hear about your results. And, as always, we're happy to help if you hit some problems; feel free to file more issues. 😄
